src/main/java/com/amazonaws/services/kinesis/leases/impl/LeaseCleanupManager.java

```diff
@@ -0,0 +1,340 @@
+
+        try {
+            if (cleanupLeasesUponShardCompletion && timeToCheckForCompletedShard) {
+                final KinesisClientLease leaseFromDDB = leaseManager.getLease(shardInfo.getShardId());
+                if(leaseFromDDB != null) {
+                    Set<String> childShardKeys = leaseFromDDB.getChildShardIds();
+                    if (CollectionUtils.isNullOrEmpty(childShardKeys)) {
+                        try {
+                            childShardKeys = getChildShardsFromService(shardInfo);
+
+                            if (CollectionUtils.isNullOrEmpty(childShardKeys)) {
+                                LOG.error("No child shards returned from service for shard " + shardInfo.getShardId());
+                            } else {
+                                wereChildShardsPresent = true;
+                                updateLeaseWithChildShards(leasePendingDeletion, childShardKeys);
+                            }
+                        } catch (ResourceNotFoundException e) {
+                            throw e;
+                        } finally {
+                            alreadyCheckedForGarbageCollection = true;
+                        }
+                    } else {
+                        wereChildShardsPresent = true;
+                    }
+                    try {
+                        cleanedUpCompletedLease = cleanupLeaseForCompletedShard(lease, childShardKeys);
+                    } catch (Exception e) {
+                        // Suppressing the exception here, so that we can attempt for garbage cleanup.
+                        LOG.warn("Unable to cleanup lease for shard " + shardInfo.getShardId());
+                    }
+                } else {
+                    LOG.info("Lease not present in lease table while cleaning the shard " + shardInfo.getShardId());
+                    cleanedUpCompletedLease = true;
+                }
+            }
+
+            if (!alreadyCheckedForGarbageCollection && timeToCheckForGarbageShard) {
+                try {
+                    wereChildShardsPresent = !CollectionUtils
+                            .isNullOrEmpty(getChildShardsFromService(shardInfo));
+                } catch (ResourceNotFoundException e) {
+                    throw e;
+                }
+            }
+        } catch (ResourceNotFoundException e) {
+            wasResourceNotFound = true;
+            cleanedUpGarbageLease = cleanupLeaseForGarbageShard(lease);
+        }
+
+        return new LeaseCleanupResult(cleanedUpCompletedLease, cleanedUpGarbageLease, wereChildShardsPresent,
+                wasResourceNotFound);
+    }
+
+    private Set<String> getChildShardsFromService(ShardInfo shardInfo) {
+        final String iterator = kinesisProxy.getIterator(shardInfo.getShardId(), ShardIteratorType.LATEST.toString());
+        return kinesisProxy.get(iterator, MAX_RECORDS).getChildShards().stream().map(c -> c.getShardId()).collect(Collectors.toSet());
```

Comment:
Let's not use MAX_RECORDS as 1. Let's stick with default.

Reply:
Will remove this limit